src/content/marker-timing.js

```diff
@@ -0,0 +1,132 @@
+ * out into multiple rows, with the overlapping timings going in the next rows. The
+ * getMarkerTiming tests show the behavior of how this works in practice.
+ *
+ * This structure allows the markers to easily be laid out like this example below:
+ *    ____________________________________________
+ *   | GC           | *--*       *--*        *--* |
+ *   |              |                             |
+ *   | Scripts      | *---------------------*     |
+ *   |              |                             |
+ *   | User Timings |    *----------------*       |
+ *   | User Timings |       *------------*        |
+ *   | User Timings |       *--*     *---*        |
+ *   |______________|_____________________________|
+ */
+export function getMarkerTiming(
+  tracingMarkers: TracingMarker[]
+): MarkerTimingRows {
+  // Each marker type will have it's own timing information, later collapse these into
+  // a single array.
+  const markerTimingsMap: Map<string, MarkerTiming[]> = new Map();
+
+  // Go through all of the markers.
+  for (let tracingMarkerIndex = 0; tracingMarkerIndex < tracingMarkers.length; tracingMarkerIndex++) {
+    const marker = tracingMarkers[tracingMarkerIndex];
+    let markerTimingsByName = markerTimingsMap.get(marker.name);
+    if (markerTimingsByName === undefined) {
+      markerTimingsByName = [];
+      markerTimingsMap.set(marker.name, markerTimingsByName);
+    }
+
+    // Place the marker in the closest row that is empty.
+    for (let i = 0; i < MAX_STACKING_DEPTH; i++) {
+      // Get or create a row for marker timings.
+      let markerTimingsRow = markerTimingsByName[i];
+      if (!markerTimingsRow) {
+        markerTimingsRow = {
+          start: [],
+          end: [],
+          index: [],
+          label: [],
+          name: marker.name,
+          length: 0,
+        };
+        markerTimingsByName.push(markerTimingsRow);
+      }
+
+      let continueSearching = false;
+
+      // Search for a spot not already taken up by another marker of this type.
+      for (let j = 0; j < markerTimingsRow.length; j++) {
+        const otherStart = markerTimingsRow.start[j];
+        const otherEnd = markerTimingsRow.end[j];
+        if (otherStart > marker.start + marker.dur) {
+          break;
+        }
+        if (otherEnd > marker.start) {
+          continueSearching = true;
+          break;
+        }
+      }
+
+      if (!continueSearching) {
+        // An empty spot was found, fill the values in the table.
+        markerTimingsRow.start.push(marker.start);
+        markerTimingsRow.end.push(marker.start + marker.dur);
+        markerTimingsRow.label.push(computeMarkerLabel(marker.data));
+        markerTimingsRow.index.push(tracingMarkerIndex);
+        markerTimingsRow.length++;
+        break;
+      }
+    }
+  }
+
+  // Flatten out the map into an array.
+  let markerTimingRows = [];
+  for (const [, value] of markerTimingsMap) {
+    markerTimingRows = markerTimingRows.concat(value);
+  }
```

Comment:
I think this just work to flatten: `[].concat(...marketTimingRows.values());`
without intermediary arrays :)